### PR TITLE
iOS Style Guide Changes

### PR DESCRIPTION
**Pull request by mattbarker016**
This improves the structure and usage of style components across the app, in line with the [new style guide](https://www.figma.com/file/QyUBYdJeesuBiphl0I3uIz1j/iOS-Design-Kit?node-id=0%3A1).

This is a **phat** PR, but the meat is in **Styles.swift**, where I declare and define the changes that I then just update throughout the app. Take a gander at this file's contents below for review.

```
/// App colors
struct Colors {
    
    // MARK: - Brand Color
    static let tcatBlue = UIColor(hex: "079DDC")
    
    // MARK: - Accent Colors
    static let lateRed = UIColor(hex: "D6304F")
    static let liveGreen = UIColor(hex: "27AE60")
    static let warningOrange = UIColor(hex: "E79C20")
    
    // MARK: - Grayscale Colors
    static let primaryText = UIColor(hex: "212121")
    static let secondaryText = UIColor(hex: "616161")
    static let metadataIcon = UIColor(hex: "9E9E9E")
    static let dividerTextField = UIColor(hex: "EEEEEE")
    static let backgroundWash = UIColor(hex: "F5F5F5")
    
    // MARK: - Constants
    static let black = UIColor.black
    static let white = UIColor.white
    
}

/// Font identifiers
struct Fonts {
    
    struct SanFrancisco {
        
        // General Sizes
        static let regular = "regular"
        static let medium = "medium"
        static let semibold = "semibold"
        static let bold = "bold"
        
        struct ProDisplay {
            static let regular = "SFProDisplay-Regular"
            static let medium = "SFProDisplay-Medium"
            static let semibold = "SFProDisplay-Semibold"
            static let bold = "SFProDisplay-Bold"
        }
        
        struct ProText {
            static let regular = "SFUIText-Regular"
            static let medium = "SFUIText-Medium"
            static let semibold = "SFUIText-Semibold"
            static let bold = "SFUIText-Bold"
        }
        
    }
    
    struct System {
        // Placeholder for init function below, not actually used.
        static let systemFontSize: CGFloat = 14 // Same as UIFont.systemFontSize
        
        static let regular = UIFont.systemFont(ofSize: systemFontSize, weight: .regular).fontName
        static let semibold = UIFont.systemFont(ofSize: systemFontSize, weight: .semibold).fontName
        static let bold = UIFont.systemFont(ofSize: systemFontSize, weight: .bold).fontName
    }
    
}

extension UIFont {
    
    /// Generate fonts for app usage
    static func style(_ name: String, size: CGFloat) -> UIFont {
        var name = name
        if size >= 14 {
            switch name {
            case Fonts.SanFrancisco.regular: name = Fonts.SanFrancisco.ProDisplay.regular
            case Fonts.SanFrancisco.medium: name = Fonts.SanFrancisco.ProDisplay.medium
            case Fonts.SanFrancisco.semibold: name = Fonts.SanFrancisco.ProDisplay.semibold
            case Fonts.SanFrancisco.bold: name = Fonts.SanFrancisco.ProDisplay.bold
            default: break
            }
        } else {
            switch name {
            case Fonts.SanFrancisco.regular: name = Fonts.SanFrancisco.ProText.regular
            case Fonts.SanFrancisco.medium: name = Fonts.SanFrancisco.ProText.medium
            case Fonts.SanFrancisco.semibold: name = Fonts.SanFrancisco.ProText.semibold
            case Fonts.SanFrancisco.bold: name = Fonts.SanFrancisco.ProText.bold
            default: break
            }
        }
        return UIFont(name: name, size: size)!
    }
    
}

struct Spacing {
    
    static let eight: CGFloat = 8
    static let twelve: CGFloat = 12
    static let fourteen: CGFloat = 14
    static let twentyFour: CGFloat = 24
    
}
```

### CODE REVIEW

**Review by Omarrasheed — APPROVED** [review]

File: TCAT/Utilities/Styles.swift
Comment (myo3) [inline TCAT/Utilities/Styles.swift]: can we call this method `font` or `getFont` instead of `style` since function returns a `UIFont`

Comment (myo3) [reply TCAT/Utilities/Styles.swift]: also think it reads better if have "name:" argument show up as a parameter label when use function (i.e. `name:` instead of `_ name:`)

File: TCAT/Utilities/Styles.swift
Comment (myo3) [inline TCAT/Utilities/Styles.swift:66]: don't switch on Strings b/c it assumes whoever uses the method will always input a valid font `name`, but if they don't, the return statement of `UIFont(name: name, size: size)!` will fail. 

`name` should be an enum and therefore `regular`, `medium`, `semibold`, and `bold` should be enums.

Also why is the `regular` font name called "regular"--is that the default system font or is it San Francisco?

Basically talking about an enum like this:
```
enum System: String {
    case regular, medium, semibold, bold
}
```
And then the `enum System` would be either within the `struct SanFrancisco`  if regular is a San Francisco font or within `struct Fonts` if regular is a default system font.

And then you can use the `rawValue` of the `enum` to get the `regular` string like:
```
let name = System.regular.rawValue // name is "regular"
```

File: TCAT/Utilities/Styles.swift
Comment (myo3) [inline TCAT/Utilities/Styles.swift:85]: idt this is necessary. unless our style guide is such that our spacing is consistent within views and across app idt we n2 have values like `eight`, we can just put in `8` for the spacing constraint.

also if we were to do spacing i think it would be better to abstract the `Spacing` in terms of the content and not number. E.g. smtg like this
```
struct Spacing { 
     static let cellTopAndBottomPadding: CGFloat = 8
}
```

But to do that would be intensive b/c we would have to name spacing for every single element and so the list would get quite long. 

Overall think its better to keep spacing as static constants within the view file where the spacing is used

File: TCAT/Views/BusIcon.swift
Comment (myo3) [inline TCAT/Views/BusIcon.swift]: maybe im missing smtg here but aren't u supposed to do: `Fonts. SanFrancisco.style(Fonts.SanFrancisco.semibold, size: fontSize)`?

**Review by myo3 — APPROVED** [review]
nice to see our style guide being updated!

File: TCAT/Views/BusIcon.swift
Comment (mattbarker016) [reply TCAT/Views/BusIcon.swift]: The function is UIFont.style, I'm just using type inference

File: TCAT/Utilities/Styles.swift
Comment (mattbarker016) [reply TCAT/Utilities/Styles.swift:85]: That makes sense. But perhaps we could define the constants in this place, and use Spacing.small in each file and give it a better name description. That way if design wanted to double whitespace, we could do so more easily.

File: TCAT/Utilities/Styles.swift
Comment (mattbarker016) [reply TCAT/Utilities/Styles.swift]: The function is technically UIFont.style. I figured since the name will always have the word font in it, it didn't need to be in the function name. 

I used `_ name:` because I figured it would be obvious what the parameter means, and would save space. Do you feel strongly about this? I don't.

File: TCAT/Utilities/Styles.swift
Comment (mattbarker016) [reply TCAT/Utilities/Styles.swift:66]: Great idea, thanks!

File: TCAT/Utilities/Styles.swift
Comment (mattbarker016) [reply TCAT/Utilities/Styles.swift]: I changed the name, kept the first param

**Comment by mattbarker016** [comment]
@myo3 The new Styles.swift is simplified down on fonts. See the changes below. I'll wait for one last sound off before I merge.

```
/// Font identifiers
enum Fonts {
    
    case regular, medium, semibold, bold
    
    enum SanFrancisco {
        
        struct ProDisplay {
            static let regular = "SFProDisplay-Regular"
            static let medium = "SFProDisplay-Medium"
            static let semibold = "SFProDisplay-Semibold"
            static let bold = "SFProDisplay-Bold"
        }
        
        struct ProText {
            static let regular = "SFUIText-Regular"
            static let medium = "SFUIText-Medium"
            static let semibold = "SFUIText-Semibold"
            static let bold = "SFUIText-Bold"
        }
        
    }
    
}

extension UIFont {
    
    /// Generate fonts for app usage
    static func getFont(_ name: Fonts, size: CGFloat) -> UIFont {
        var fontString: String
        if size >= 14 {
            switch name {
            case .regular: fontString = Fonts.SanFrancisco.ProDisplay.regular
            case .medium: fontString = Fonts.SanFrancisco.ProDisplay.medium
            case .semibold: fontString = Fonts.SanFrancisco.ProDisplay.semibold
            case .bold: fontString = Fonts.SanFrancisco.ProDisplay.bold
            }
        } else {
            switch name {
            case .regular: fontString = Fonts.SanFrancisco.ProText.regular
            case .medium: fontString = Fonts.SanFrancisco.ProText.medium
            case .semibold: fontString = Fonts.SanFrancisco.ProText.semibold
            case .bold: fontString = Fonts.SanFrancisco.ProText.bold
            }
        }
        return UIFont(name: fontString, size: size)!
    }
    
}
```

**Pull request merged**
